samples/amber.cc

```diff
@@ -178,43 +182,88 @@ int main(int argc, const char** argv) {
     return 0;
   }
 
-  if (options.input_filename.empty()) {
+  if (options.input_filenames.empty()) {
     std::cerr << "Input file must be provided." << std::endl;
     return 2;
   }
 
-  auto data = ReadFile(options.input_filename);
-  if (data.empty())
-    return 1;
+  amber::Result result;
+  std::vector<std::string> failures;
+  std::vector<std::pair<std::unique_ptr<amber::Recipe>, std::string>>
+      recipe_and_files;
+  for (const auto& file : options.input_filenames) {
+    auto data = ReadFile(file);
+    if (data.empty()) {
+      std::cerr << file << " is empty." << std::endl;
+      failures.push_back(file);
+      continue;
+    }
 
-  amber::Amber am;
-  amber::Recipe recipe;
-  amber::Result result = am.Parse(data, &recipe);
-  if (!result.IsSuccess()) {
-    std::cerr << result.Error() << std::endl;
-    return 1;
+    amber::Amber am;
+    std::unique_ptr<amber::Recipe> recipe = amber::MakeUnique<amber::Recipe>();
+
+    result = am.Parse(data, recipe.get());
+    if (!result.IsSuccess()) {
+      std::cerr << "case " << file << ": parse fail" << std::endl;
+      std::cerr << result.Error() << std::endl << std::endl;
+      failures.push_back(file);
+      continue;
+    }
+
+    recipe_and_files.push_back(
+        std::make_pair(std::move(recipe), std::string(file)));
   }
 
   if (options.parse_only)
     return 0;
 
+  sample::ConfigHelper config_helper;
   amber::Options amber_options;
   amber_options.engine = options.engine;
-  result = am.Execute(&recipe, amber_options);
-  if (!result.IsSuccess()) {
-    std::cerr << result.Error() << std::endl;
-    return 1;
-  }
+  if (options.input_filenames.size() > 1UL) {
```

Comment:
This check isn't needed, as you may have less things in recipes then in input_files depending on if the parse failed.

Reply:
Always create the device externally, we don't need to special case if there is a single filename.